Auth0/SafariAuthenticationSession.swift

```diff
@@ -22,31 +22,45 @@
 
 import UIKit
 import SafariServices
+import AuthenticationServices
 
 #if swift(>=3.2)
 @available(iOS 11.0, *)
 class SafariAuthenticationSession: AuthSession {
 
-    typealias FinishSession = (Result<Credentials>) -> Void
-
-    var authSession: SFAuthenticationSession?
-    let authorizeURL: URL
+    private var authSession: NSObject?
 
     init(authorizeURL: URL, redirectURL: URL, state: String? = nil, handler: OAuth2Grant, finish: @escaping FinishSession, logger: Logger?) {
-        self.authorizeURL = authorizeURL
         super.init(redirectURL: redirectURL, state: state, handler: handler, finish: finish, logger: logger)
-        self.authSession = SFAuthenticationSession(url: self.authorizeURL, callbackURLScheme: self.redirectURL.absoluteString) { [unowned self] in
-            guard $1 == nil, let callbackURL = $0 else {
-                if case SFAuthenticationError.canceledLogin = $1! {
-                    self.finish(.failure(error: WebAuthError.userCancelled))
-                } else {
-                    self.finish(.failure(error: $1!))
+        if #available(iOS 12.0, *) {
+            let authSession = ASWebAuthenticationSession(url: authorizeURL, callbackURLScheme: self.redirectURL.absoluteString) { [unowned self] in
```

Comment:
Seems like a lot of duplication of code in side these blocks, could we refactor this down to one block and a typealias?